### DISCUSSION TITLE
Generic Overload for `Assert.DoesNotThrow`

### DISCUSSION THREAD

**Discussion by Diemaco** [nunit/nunit#4395] [Ideas]
### Problem
NUnit 3 currently provides `Assert.Throws` and `Assert.Throws<TActual>` methods to verify that specific exceptions are thrown during test execution. However, it lacks a generic overload for `Assert.DoesNotThrow`.

### Solution
A generic overload of the `Assert.DoesNotThrow` method, for example `Assert.DoesNotThrow<TActual>`.

### Example Usage 
```csharp
Assert.DoesNotThrow<InvalidOperationException>(() => myObject.MyMethod());
```

**Comment by Diemaco** [comment]
Stack Overflow question about the same thing: [Where is the generic DoesNotThrow<T> method analogous to the Throws<T> method?](https://stackoverflow.com/questions/23041293/where-is-the-generic-doesnotthrowt-method-analogous-to-the-throwst-method)

**Comment by manfred-brands** [comment]
What is the practical use case for this?

The Stack Overflow question says:
> Currently I just use the DoesNotThrow version that's available, and pray it will never hide exceptions of a type that does indicate a problem.

This will never hide exceptions, as it will complain if any exceptions are thrown.
Even the accepted answer will simply always fail if any exception is thrown.

If you use `Assert.DoesNotThrow<InvalidOperationException>` with a specific exception does that mean you are happy if it throws another one? 
In theory the whole method can be omitted as the test would fail if any exception is thrown, regardless if you use `DoesNotThrow` or not.

**Reply by Diemaco** [reply]
- With `Assert.DoesNotThrow<T>`, you can specify why the tested method shouldn't throw exception T with a message, which should aid in debugging.
- If you don't specify a particular exception type, the test will still fail if any exception is thrown. However, using the generic overload allows you to communicate your intent better.
